source/h2_connection.c

```diff
@@ -1111,6 +1120,51 @@ static struct aws_h2err s_decoder_on_window_update(uint32_t stream_id, uint32_t
     return AWS_H2ERR_SUCCESS;
 }
 
+struct aws_h2err s_decoder_on_goaway_begin(
+    uint32_t last_stream,
+    uint32_t error_code,
+    uint32_t debug_data_length,
+    void *userdata) {
+    (void)debug_data_length;
+    struct aws_h2_connection *connection = userdata;
+
+    if (last_stream > connection->thread_data.goaway_received_last_stream_id) {
+        CONNECTION_LOGF(
+            ERROR,
+            connection,
+            "Peer GOAWAY frame has invalid last stream id %" PRIu32 ", which is higher than the one we received",
+            last_stream);
+        return aws_h2err_from_h2_code(AWS_H2_ERR_PROTOCOL_ERROR);
+    }
+    /* stop sending any new stream and making new request */
+    aws_atomic_store_int(&connection->synced_data.new_stream_error_code, AWS_ERROR_HTTP_GOAWAY_RECEIVED);
+    connection->thread_data.goaway_received = true;
+    connection->thread_data.goaway_received_error_code = error_code;
+    connection->thread_data.goaway_received_last_stream_id = last_stream;
+    CONNECTION_LOGF(
+        DEBUG,
+        connection,
+        "Peer sent a GOAWAY frame, with error code %" PRIu32 ". New streams will not be accepted.",
+        error_code);
+    /* Complete activated streams whose id is higher than last_stream, since they will not process by peer. We should
+     * treat them as they had never been created at all */
+    struct aws_hash_iter stream_iter = aws_hash_iter_begin(&connection->thread_data.active_streams_map);
+    while (!aws_hash_iter_done(&stream_iter)) {
```

Comment:
Another way will be storing the highest ID we processed, iterating the ID instead of the hash map. I'm not sure which one will be better for real life.